README
Carp-REPL 0.09

Carp::REPL is a drop-in module for getting a REPL (read-eval-print-loop)
which helps you in examining the state of the world when things blew up.
Essentially, you just add -MCarp::REPL to your script's invocation. One
advantage of using Carp::REPL over Devel::REPL is the lexical variables are
made available, as well as stack manipulation commands.

MAJOR CHANGES
    0.08: 'nodie' and 'warn'
    0.07: sub arguments stuffed into $_a
    0.06: proper aliases so you can influence the environment
    0.05: :bt to redisplay stack trace, :q to quit
    0.03: :up and :down for traversing the call stack
    0.01: first version

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the perldoc command.

    perldoc Carp::REPL

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/Carp-REPL

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Carp-REPL

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/Carp-REPL

    CPAN Ratings:
        http://cpanratings.perl.org/d/Carp-REPL

COPYRIGHT AND LICENCE

Copyright (C) 2007 Best Practical Solutions

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for Carp-REPL

0.09    Thu Jul 12 12:32:12 2007
        Move to Module::Install

0.08    Wed Jul 11 16:56:08 2007
        Add 'nodie' and 'warn' arguments

0.07    Tue Jul 10 22:34:55 2007
        Made sub arguments available in $_a

0.06    Sat Jul  7 13:43:14 EDT 2007
        Proper aliasing support
        Cleaned up the internals

0.05    Fri Jul  6 18:47:33 EDT 2007
        :bt command to redisplay the backtrace
        REPL now has History plugin loaded
        Various bug fixes
        :q command

0.04    Fri Jul  6 14:57:33 EDT 2007
        Fix annoying global variable issue.
        But now REPL doesn't have use strict 'vars'.
        Will be fixed as soon as my Devel::EvalContext patch is applied.

0.03    Fri Jul  6 14:17:46 EDT 2007
        Add ability to move up and down the stack (:u and :d).
        Moose startup penalty now only paid if you get a REPL.

0.02    Fri Jul  6 01:04:33 EDT 2007
        Add the somewhat bloaty dependencies to Makefile.PL.

0.01    Fri Jul  6 00:35:37 EDT 2007
        First version, released on an unsuspecting world.



lib/Carp/REPL.pm
package Carp::REPL;
use strict;
use warnings;

sub import
{
    my $nodie = grep {$_ eq 'nodie'} @_;
    my $warn  = grep {$_ eq 'warn' } @_;

    $SIG{__DIE__} = \&repl unless $nodie;
    $SIG{__WARN__} = \&repl if $warn;
}

=head1 NAME

Carp::REPL - read-eval-print-loop on die

=head1 VERSION

Version 0.09 released 12 Jul 07

=cut

our $VERSION = '0.09';

=head1 SYNOPSIS

The intended way to use this module is through the command line.

    perl tps-report.pl
        Can't call method "cover_sheet" without a package or object reference at tps-report.pl line 6019.

    perl -MCarp::REPL tps-report.pl
        Can't call method "cover_sheet" without a package or object reference at tps-report.pl line 6019.

        $ map {"$_\n"} $form, $subform
        27B/6
        Report::TPS::Subreport=HASH(0x86da61c)

=head1 USAGE

    -MCarp::REPL

Works as command line argument. This automatically installs the die handler for
you, so if you receive a fatal error you get a REPL before the universe
explodes.

    use Carp::REPL;

Same as above.

    use Carp::REPL 'nodie';

Loads the module without installing the die handler. Use this if you just want to
run C<Carp::REPL::repl> on your own terms.

    use Carp::REPL 'warn';

Same as C<Carp::REPL> but also installs REPL to be invoked whenever a warning
is generated.

    use Carp::REPL 'warn', 'nodie';

I don't see why you would want to do this, but it's available. :)

=head1 FUNCTIONS

=head2 repl

This module's interface consists of exactly one function: repl. This is
provided so you may install your own C<$SIG{__DIE__}> handler if you have no
alternatives.

It takes the same arguments as die, and returns no useful value. In fact, don't
even depend on it returning at all!

One useful place for calling this manually is if you just want to check the
state of things without having to throw a fake error. You can also change any
variables and those changes will be seen by the rest of your program.

    use Carp::REPL;

    sub involved_calculation
    {
        # ...
        $d = maybe_zero();
        # ...
        Carp::REPL::repl; # $d = 1
        $sum += $n / $d;
        # ...
    }

Unfortunately if you instead go with the usual C<-MCarp::REPL>, then
C<$SIG{__DIE__}> will be invoked and there's no general way to recover. But you
can still change variables and poke at things.

=cut

sub repl
{
    warn @_, "\n"; # tell the user what blew up

    require PadWalker;
    require Devel::REPL;

    my (@packages, @environments, @argses, $backtrace);

    my $frame = 0;
    while (1)
    {
        package DB;
        my ($package, $file, $line, $subroutine) = caller($frame)
            or last;
        $package = 'main' if !defined($package);

        eval
        {
            # PadWalker has 0 mean 'current'
            # caller has 0 mean 'immediate caller'
            push @environments, PadWalker::peek_my($frame+1);
        };
        Carp::carp($@), last if $@;

        push @argses, [@DB::args];
        push @packages, [$package, $file, $line];

        $backtrace .= sprintf "%s%d: %s called at %s:%s.\n",
            $frame == 0 ? '' : '   ',
            $frame,
            $subroutine,
            $file,
            $line;
        ++$frame;
    }

    warn $backtrace;

    my $repl = Devel::REPL->new;

    # LexEnv must come before LexEnvCarp
    $repl->load_plugin($_) for qw/History LexEnv LexEnvCarp/;

    $repl->environments(\@environments);
    $repl->packages(\@packages);
    $repl->argses(\@argses);
    $repl->backtrace($backtrace);
    $repl->frame(0);

    $repl->run;
}

=head1 COMMANDS

Note that this is not supposed to be a full-fledged debugger. A few commands
are provided to aid you in finding out what went awry. See
L<Devel::ebug> if you're looking for a serious debugger.

=over 4

=item * :u

Moves one frame up in the stack.

=item * :d

Moves one frame down in the stack.

=item * :t

Redisplay the stack trace.

=item * :q

Close the REPL. (C<^D> also works)

=back

=head1 VARIABLES

=over 4

=item * $_REPL

This represents the Devel::REPL object (with the LexEnvCarp plugin, among
others, mixed in).

=item * $_a

This represents the arguments passed to the subroutine at the current frame in
the call stack. Modifications are ignored (how would that work anyway?
Re-invoke the sub?)

=back

=head1 CAVEATS

Dynamic scope probably produces unexpected results. I don't see any easy (or
even difficult!) solution to this. Therefore it's a caveat and not a bug. :)

=head1 SEE ALSO

L<Devel::REPL>, L<Devel::ebug>

=head1 AUTHOR

Shawn M Moore, C<< <sartak at gmail.com> >>

=head1 BUGS

No known bugs at this point. To expect that to stay true is laughably naive. :)

Please report any bugs or feature requests to
C<bug-carp-repl at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Carp-REPL>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Carp::REPL

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Carp-REPL>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Carp-REPL>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Carp-REPL>

=item * Search CPAN

L<http://search.cpan.org/dist/Carp-REPL>

=back

=head1 ACKNOWLEDGEMENTS

Thanks to Nelson Elhage and Jesse Vincent for the idea.

Thanks to Matt Trout and Stevan Little for their advice.

=head1 COPYRIGHT & LICENSE

Copyright 2007 Best Practical Solutions, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Carp::REPL



lib/Devel/REPL/Plugin/LexEnvCarp.pm
package Devel::REPL::Plugin::LexEnvCarp;

use Moose::Role;
use namespace::clean -except => [ 'meta' ];
use Devel::LexAlias;

has 'environments' => (
    isa => 'ArrayRef',
    is => 'rw',
    required => 1,
    default => sub { [{}] },
);

has 'packages' => (
    isa => 'ArrayRef',
    is => 'rw',
    required => 1,
    default => sub { ['main'] },
);

has 'argses' => (
    isa => 'ArrayRef',
    is => 'rw',
    required => 1,
    default => sub { [[]] },
);

has 'frame' => (
    isa => 'Int',
    is => 'rw',
    required => 1,
    default => 0,
);

has 'backtrace' => (
    isa => 'Str',
    is => 'rw',
    required => 1,
    default => '',
);

around 'frame' => sub
{
    my $orig = shift;

    my ($self, $frame) = @_;

    return $orig->(@_) if !defined($frame);

    if ($frame < 0)
    {
        warn "You're already at the bottom frame.\n";
    }
    elsif ($frame >= @{ $self->packages })
    {
        warn "You're already at the top frame.\n";
    }
    else
    {
        my ($package, $file, $line) = @{$self->packages->[$frame]};
        print "Now at $file:$line (frame $frame).\n";
        $orig->(@_);
    }
};

# this is totally the wrong spot for this. oh well.
around 'read' => sub
{
  my $orig = shift;
  my ($self, @rest) = @_;
  my $line = $self->$orig(@rest);

  return if !defined($line) || $line =~ /^\s*:q\s*$/;

  if ($line =~ /^\s*:b?t\s*$/)
  {
    print $self->backtrace;
    return '';
  }

  if ($line =~ /^\s*:up?\s*$/)
  {
    $self->frame($self->frame + 1);
    return '';
  }

  if ($line =~ /^\s*:d(?:own)?\s*$/)
  {
    $self->frame($self->frame - 1);
    return '';
  }

  return $line;
};

around 'mangle_line' => sub
{
  my $orig = shift;
  my ($self, @rest) = @_;
  my $line = $self->$orig(@rest);

  my $frame = $self->frame;
  my $package = $self->packages->[$frame][0];

  my $declarations = join "\n",
                     map {"my $_;"}
                     keys %{ $self->environments->[$frame] };

  my $aliases = << 'ALIASES';
while (my ($k, $v) = each %{ $_REPL->environments->[$_REPL->frame] })
{
    Devel::LexAlias::lexalias 0, $k, $v;
}
my $_a; Devel::LexAlias::lexalias 0, '$_a', \$_REPL->argses->[$_REPL->frame];
ALIASES

  return << "CODE";
package $package;
no warnings 'misc'; # declaration in same scope masks earlier instance
no strict 'vars';   # so we get all the global variables in our package
$declarations
$aliases
$line
CODE
};

=head1 NAME

Devel::REPL::Plugin::LexEnvCarp - Devel::REPL plugin for Carp::REPL

=head1 VERSION

Version 0.09 released 12 Jul 07

=cut

our $VERSION = '0.09';

=head1 SYNOPSIS

This sets up the environment captured by L<Carp::REPL>. This plugin
isn't intended for use by anything else. There are plans to move some features
from this into a generic L<Devel::REPL> plugin.

This plugin also adds a few extra commands like :up and :down to move up and
down the stack.

=head1 AUTHOR

Shawn M Moore, C<< <sartak at gmail.com> >>

=head1 BUGS

Please report any bugs to a medium given by Carp::REPL.

=head1 COPYRIGHT & LICENSE

Copyright 2007 Best Practical Solutions, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of Devel::REPL::Plugin::LexEnvCarp
